src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs

```diff
@@ -0,0 +1,1564 @@
+            {
+                if (attribute.Target?.GetAttributeLocation() == AttributeLocation.Field)
+                {
+                    diagnostics.Add(
+                        new CSDiagnosticInfo(ErrorCode.WRN_AttributesOnBackingFieldsNotAvailable,
+                            languageVersion.ToDisplayString(),
+                            new CSharpRequiredLanguageVersion(MessageID.IDS_FeatureAttributesOnBackingFields.RequiredVersion())),
+                        attribute.Target.Location);
+                }
+            }
+        }
+
+        protected abstract void GetAccessorDeclarations(
+            CSharpSyntaxNode syntax,
+            DiagnosticBag diagnostics,
+            out bool isAutoProperty,
+            out bool hasAccessorList,
+            out bool accessorsHaveImplementation,
+            out bool hasGetAccessor,
+            out bool hasSetAccessor,
+            out bool isInitOnly,
+            out CSharpSyntaxNode? getSyntax,
+            out CSharpSyntaxNode? setSyntax);
+
+        protected abstract void CheckForBlockAndExpressionBody(CSharpSyntaxNode syntax, DiagnosticBag diagnostics);
+
+#nullable restore
+
+        internal sealed override ImmutableArray<string> NotNullMembers =>
+            GetDecodedWellKnownAttributeData()?.NotNullMembers ?? ImmutableArray<string>.Empty;
+
+        internal sealed override ImmutableArray<string> NotNullWhenTrueMembers =>
+            GetDecodedWellKnownAttributeData()?.NotNullWhenTrueMembers ?? ImmutableArray<string>.Empty;
+
+        internal sealed override ImmutableArray<string> NotNullWhenFalseMembers =>
+            GetDecodedWellKnownAttributeData()?.NotNullWhenFalseMembers ?? ImmutableArray<string>.Empty;
+
+        internal bool IsExpressionBodied
+            => (_propertyFlags & Flags.IsExpressionBodied) != 0;
+
+        private void CheckInitializer(
+            bool isAutoProperty,
+            bool isInterface,
+            bool isStatic,
+            Location location,
+            DiagnosticBag diagnostics)
+        {
+            if (isInterface && !isStatic)
+            {
+                diagnostics.Add(ErrorCode.ERR_InstancePropertyInitializerInInterface, location, this);
+            }
+            else if (!isAutoProperty)
+            {
+                diagnostics.Add(ErrorCode.ERR_InitializerOnNonAutoProperty, location, this);
+            }
+        }
+
+        public override RefKind RefKind
+        {
+            get
+            {
+                return _refKind;
+            }
+        }
+
+        public override TypeWithAnnotations TypeWithAnnotations
+        {
+            get
+            {
+                if (_lazyType == null)
+                {
+                    var diagnostics = DiagnosticBag.GetInstance();
+                    var binder = this.CreateBinderForTypeAndParameters();
+                    var result = this.ComputeType(binder, CSharpSyntaxNode, diagnostics);
```

Comment:
>var result = this.ComputeType(binder, CSharpSyntaxNode, diagnostics); [](start = 20, length = 69)

I think the process of calculating the type should be abstracted. For example we don't want to bind type again in same cases. For example, when a property is created based on record parameter.

Reply:
To clarify. I think it is fine to keep ComputeType and ComputeParameters helpers in this type, but this type shouldn't call them directly, derived types can call them when there is really a syntax to bind, otherwise a derived type can do something else

---
In reply to: [445208055](https://github.com/dotnet/roslyn/pull/45411#discussion_r445208055) [](ancestors = 445208055)